### PR TITLE
Add NumPy implementation

### PR DESCRIPTION
**Pull request by gordicaleksa**
Added a NumPy implementation closely following the style of PyTorch reference implementation.

There are some minor differences when doing the weight update:
```
    grads = model.backward()
    optimizer.step(grads)
```
vs PyTorch:
```
    loss.backward()
    optimizer.step()
    optimizer.zero_grad()
```
can be fixed if we want to build the underlying graph, similar to PyTorch, during the fwd. Will implement if this was the idea.

Tested for equivalency -> logits, loss, grads are all nearly == during the training.

Due to a different rng implementation in NumPy/PyTorch we can't just set the same seed and get the same init weights even if we implement the same init methods (kaiming, normal, etc.). So for now i just save the init weights in PyTorch reference and use those to init the NumPy network - open for suggestions here!

First 4k steps in PyTorch:
```
step 0/50000 | train_loss 3.3584 | val_loss 3.3592 | lr 1.000000e-03
step 200/50000 | train_loss 2.4595 | val_loss 2.4625 | lr 9.999605e-04
step 400/50000 | train_loss 2.3869 | val_loss 2.4026 | lr 9.998421e-04
step 600/50000 | train_loss 2.3637 | val_loss 2.3698 | lr 9.996447e-04
step 800/50000 | train_loss 2.3619 | val_loss 2.3574 | lr 9.993685e-04
step 1000/50000 | train_loss 2.3629 | val_loss 2.3538 | lr 9.990134e-04
step 1200/50000 | train_loss 2.3360 | val_loss 2.3390 | lr 9.985795e-04
step 1400/50000 | train_loss 2.3284 | val_loss 2.3178 | lr 9.980668e-04
step 1600/50000 | train_loss 2.3046 | val_loss 2.3103 | lr 9.974755e-04
step 1800/50000 | train_loss 2.3150 | val_loss 2.3057 | lr 9.968057e-04
step 2000/50000 | train_loss 2.3071 | val_loss 2.2936 | lr 9.960574e-04
step 2200/50000 | train_loss 2.2947 | val_loss 2.2895 | lr 9.952307e-04
step 2400/50000 | train_loss 2.2837 | val_loss 2.2841 | lr 9.943259e-04
step 2600/50000 | train_loss 2.2768 | val_loss 2.2751 | lr 9.933430e-04
step 2800/50000 | train_loss 2.2791 | val_loss 2.2730 | lr 9.922822e-04
step 3000/50000 | train_loss 2.2816 | val_loss 2.2621 | lr 9.911436e-04
step 3200/50000 | train_loss 2.2881 | val_loss 2.2529 | lr 9.899275e-04
step 3400/50000 | train_loss 2.1802 | val_loss 2.2553 | lr 9.886341e-04
step 3600/50000 | train_loss 2.2363 | val_loss 2.2527 | lr 9.872634e-04
step 3800/50000 | train_loss 2.2392 | val_loss 2.2369 | lr 9.858159e-04
step 4000/50000 | train_loss 2.2385 | val_loss 2.2364 | lr 9.842916e-04
```

First 4k steps in NumPy:
```
step 0/50000 | train_loss 3.3584 | val_loss 3.3592 | lr 1.000000e-03
step 200/50000 | train_loss 2.4595 | val_loss 2.4625 | lr 9.999605e-04
step 400/50000 | train_loss 2.3869 | val_loss 2.4026 | lr 9.998421e-04
step 600/50000 | train_loss 2.3637 | val_loss 2.3698 | lr 9.996447e-04
step 800/50000 | train_loss 2.3619 | val_loss 2.3574 | lr 9.993685e-04
step 1000/50000 | train_loss 2.3630 | val_loss 2.3538 | lr 9.990134e-04
step 1200/50000 | train_loss 2.3361 | val_loss 2.3391 | lr 9.985795e-04
step 1400/50000 | train_loss 2.3285 | val_loss 2.3179 | lr 9.980668e-04
step 1600/50000 | train_loss 2.3047 | val_loss 2.3103 | lr 9.974755e-04
step 1800/50000 | train_loss 2.3152 | val_loss 2.3058 | lr 9.968057e-04
step 2000/50000 | train_loss 2.3073 | val_loss 2.2938 | lr 9.960574e-04
step 2200/50000 | train_loss 2.2949 | val_loss 2.2896 | lr 9.952307e-04
step 2400/50000 | train_loss 2.2838 | val_loss 2.2843 | lr 9.943259e-04
step 2600/50000 | train_loss 2.2771 | val_loss 2.2754 | lr 9.933430e-04
step 2800/50000 | train_loss 2.2795 | val_loss 2.2733 | lr 9.922822e-04
step 3000/50000 | train_loss 2.2821 | val_loss 2.2625 | lr 9.911436e-04
step 3200/50000 | train_loss 2.2886 | val_loss 2.2533 | lr 9.899275e-04
step 3400/50000 | train_loss 2.1801 | val_loss 2.2557 | lr 9.886341e-04
step 3600/50000 | train_loss 2.2366 | val_loss 2.2532 | lr 9.872634e-04
step 3800/50000 | train_loss 2.2395 | val_loss 2.2372 | lr 9.858159e-04
step 4000/50000 | train_loss 2.2390 | val_loss 2.2369 | lr 9.842916e-04
```

### CODE REVIEW

**Comment by gordicaleksa** [comment]
Note: we are not using `torch.no_grad` in the reference imp right now, so similarly i don't have any guards against collecting activations in the fwd step even when we don't need it inside `eval_split` func. Can easily change this in both imps.
